### PR TITLE
[Enhancement] adjust the BE and CN schedule policy

### PR DESCRIPTION
**Pull request by MatthewH00**
Why I'm doing:
In shared nothing mode，when deploy both BE and CN, and set `prefer_compute_node=true`, current BE and CN schedule policy is not very feel perfect:
1. when query internal table, except scan fragment allocate to BE, other calculate fragment would allocate to CN.
2. when query external table, all fragment would allocate to CN.

In some scenarios, like when sudden business peak, need temporarily add CN to increase computational resources, current BE and CN schedule policy would cause BE nodes have low resource usage ratio，CN nodes have high resource usage ratio，especially when business type belongs to computational tasks.

What I'm doing:
Add a new session variable `computation_fragment_scheduling_policy`, the value as follows:
1. `compute_nodes_only`: the previous policy, as the default value.
2. `all_nodes`:
    - when query internal table, scan fragment allocate to BE, other calculate fragment would allocate to both BE and CN。
    - when query external table, all fragment would allocate to both BE and CN。

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by MatthewH00** [comment]
@kevincai Hi could you please approval the workflow? the pr is from the previous(38498), i update and reopen it.

**Comment by MatthewH00** [comment]
@kevincai it looks like fail at SQL-Tester 's test_object_dependencies, Could you please check the Checker? i think the pr have little relation with object_dependencies.

**Comment by MatthewH00** [comment]
@kevincai Hi Could you please review the pr when have free time? In our prod env， we use the be/cn schedule policy in the pr，also i want listen to the community's suggestion for feasibility of the schedule policy.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java [hunk 1]
```diff
@@ -326,8 +332,26 @@ private static ImmutableMap<Long, ComputeNode> buildComputeNodeInfo(SystemInfoSe
...
             }
-            return ImmutableMap.copyOf(computeNodes);
         }
+
+        //add BE to Node Pool
+        if (computationFragmentSchedulingPolicy.equals("all_nodes")) {
+            ImmutableMap<Long, ComputeNode> idToBackend
+                    = ImmutableMap.copyOf(systemInfoService.getIdToBackend());
+            for (int i = 0; i < idToBackend.size(); i++) {
+                ComputeNode backend =
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java:342]: does the total number of compute nodes will still be enforced by taking the be as computenodes?

e.g. if the system has 2 cn and 3 be nodes and `numUsedComputeNodes` is set to `3`, it will be 2 cn nodes with one additional be node, or it will be 2 cn nodes plus all 3 be nodes?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 2]
```diff
@@ -2168,6 +2174,19 @@ public void setPreferComputeNode(boolean preferComputeNode) {
         this.preferComputeNode = preferComputeNode;
     }
 
+    public void setComputationFragmentSchedulingPolicy(String computationFragmentSchedulingPolicy) {
+        if (!"compute_nodes_only".equals(computationFragmentSchedulingPolicy)
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:2178]: define string constant variable or enum type for `compute_nodes_only`/`all_nodes`. `computation_fragment_scheduling_policy` is not easy to understand, can think of a better name for it?

**Review by kevincai — COMMENTED** [review]
overall looks good to me. need some tuning on the session variable names and tiny fixes to the selection probably?

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java (hunk 1 above)
Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java:342]: It should be 2 cn nodes plus all 3 be nodes. when face sudden business peak, current be have not enough computational power, need deploy additional cn to increace. Then total computational power should be all be and numUsedComputeNodes's cn.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 2 above)
Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:2178]: the session variable name and value of computation_fragment_scheduling_policy,  is according to ZiheLiu(github ID) 's suggestion from previous pr(#38498) , could i adjust it ?

Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:2178]: ok then, make it constant variable or enum at least, don't hardcode it everywhere.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java (hunk 1 above)
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java:342]: I don't think so.

This new session variable just controls whether need to take BE nodes as CN nodes. whether to use cn nodes, how many cn nodes to be used, will be still controlled by `preferComputeNode` and `useComputeNodes`, you can't say if use `all_nodes`, even the session variable set `preferComputeNode=false`, the compute nodes will be used anyway, that's doesn't make sense.

So in normal case
* `preferComputeNode = true`
* `useComputeNodes = -1`
* `computation_fragment_scheduling_policy = compute_nodes_only `

This will use only compute nodes all the computation works, BE will be excluded.

And when facing the sudden peak load, set the session variable
* `computation_fragment_scheduling_policy = all_nodes ` to let BE nodes working as compute nodes as well.

`preferComputeNode` and `useComputeNodes` should be always obeyed no matter what `computation_fragment_scheduling_policy` is.

Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java:342]: yes , the pr is obey the policy as you say. I offer the FE UT in DefaultWorkerProviderTest.java, the testSelectBackendAndComputeNode function display the test demo:
SystemInfoService has 5 Backend and 3 ComputeNode
- case 1.test ComputeNode only：preferComputeNode=true、useComputeNodes=[-1, 0, 2, 4]、computation_fragment_scheduling_policy=compute_nodes_only
-  case 2.test Backend only：preferComputeNode=false、useComputeNodes=[-1, 0, 2, 4]、computation_fragment_scheduling_policy=compute_nodes_only
-  case 3.test Backend and ComputeNode：preferComputeNode=true、useComputeNodes=[-1, 0, 2, 4]、computation_fragment_scheduling_policy=all_nodes

File: fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java [hunk 3]
```diff
@@ -144,6 +144,69 @@ public ImmutableMap<Long, ComputeNode> getIdComputeNode() {
...
+                            false, numUsedComputeNodes, "compute_nodes_only", WarehouseManager.DEFAULT_WAREHOUSE_ID);
+            List<Long> selectedWorkerIdsList = workerProvider.getAllAvailableNodes();
+            Assert.assertEquals(availableId2Backend.size(), selectedWorkerIdsList.size());
+            for (Long selectedWorkerId : selectedWorkerIdsList) {
+                Assert.assertTrue("selectedWorkerId:" + selectedWorkerId,
+                        availableId2Backend.containsKey(selectedWorkerId));
+            }
+        }
+        // test Backend and ComputeNode
+        for (Integer numUsedComputeNodes : numUsedComputeNodesList) {
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:192]: when numUsedComputeNode < ComputeNodes.size(), and it is the `all_nodes` policy, what's the expected behavior?

when it is `compute_nodes_only`, the candidate workers has only `numUsedComputeNode` compute nodes, when it is `all_nodes`, the candidate workers will has `numUsedComputeNode` compute nodes plus all the be nodes based on this PR's implementation, looks like not make sense?

Comment (MatthewH00) [reply fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:192]: if the user understand the `preferComputeNode/useComputeNodes/computation_fragment_scheduling_policy` three session variable 's correlation and cooperation , and the useComputeNodes is set by user self，will it make sense when the case of numUsedComputeNode < ComputeNodes.size() ?

Comment (kevincai) [reply fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:192]: maybe just don't want the query occupy too much resource? e.g. for some offline batch jobs, limit the number of used compute nodes to ETL data into sr local table.

Comment (MatthewH00) [reply fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:192]: i think make clear of `preferComputeNode/useComputeNodes/computation_fragment_scheduling_policy` three session variable 's correlation and cooperation in the user document，when the case like you say，user would understand the `all_nodes=numUsedComputeNode cn plus all be`

Comment (kevincai) [reply fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:192]: it is the scope of `computation_fragment_scheduling_policy` session variable, will it just control if take BE as CN this single point, or will it be also to overwrite the `useComputeNodes` settings?

I am opt to the former. this PR's implementation is the later, and not fully overwritten, just partially overwritten.

Comment (MatthewH00) [reply fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:192]: i will update the pr accoding to the former. the former is really clear more than later, just think about for a moment.

Comment (MatthewH00) [reply fe/fe-core/src/test/java/com/starrocks/qe/scheduler/DefaultWorkerProviderTest.java:197]: @kevincai i updated the pr to adjust the scope of computation_fragment_scheduling_policy, could you please approval the workflow ?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 4]
```diff
@@ -98,6 +98,13 @@ public class SessionVariable implements Serializable, Writable, Cloneable {
 
     public static final String USE_COMPUTE_NODES = "use_compute_nodes";
     public static final String PREFER_COMPUTE_NODE = "prefer_compute_node";
+    // The schedule policy of backend and compute nodes.
+    // The optional values are "compute_nodes_only" and "all_nodes".
+    public static final String COMPUTATION_FRAGMENT_SCHEDULING_POLICY = "computation_fragment_scheduling_policy";
+    public enum ComputationFragmentSchedulingPolicy {
+        compute_nodes_only, //only select compute node in scheduler policy (default)
+        all_nodes //both select compute node and backend in scheduler policy
+    }
```
[the remark this answers is not in the document]
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:107]: use enum LITERAL combining with string values, to interact with user's session variable, internally, string value will be converted to ComputationFragmentSchedulingPolicy object, and compared to COMPUTE_NODE_ONLY or ALL_NODES, for the validation check.

[the remark this answers is not in the document]
Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:107]: this is just an example, may or may not compile at all.

[the remark this answers is not in the document]
Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: I have updated the variable computationFragmentSchedulingPolicy to enum , could you please approve the workflow?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 5]
```diff
@@ -781,6 +801,10 @@ public static MaterializedViewRewriteMode parse(String str) {
     @VariableMgr.VarAttr(name = PREFER_COMPUTE_NODE)
     private boolean preferComputeNode = false;
 
+    @VariableMgr.VarAttr(name = COMPUTATION_FRAGMENT_SCHEDULING_POLICY)
+    private ComputationFragmentSchedulingPolicy computationFragmentSchedulingPolicy =
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: need some simple test to ensure the session variable can be set via string value. Either through MockCluster or through the sql-tester framework.

**Review by kevincai — COMMENTED** [review]
overall looks good to me.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 5 above)
Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: How to use the sql-tester framework to check it ? Could you please give some tip ?

Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: check the content under https://github.com/StarRocks/starrocks/tree/main/test. create new test cases running sql queries and validate the result.

Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: Need to validate the user set value in `SetStmtAnalyzer::analyzeSystemVariable` and add some UT in `AnalyzeSetVariableTest`.

Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: Its type should be `String`, because the type of session variables cannot be `enum`. See `VariableMgr::setValue` for detail.

Comment (ZiheLiu) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: Could refer to the implementation of the session variable `choose_execute_instances_mode`, since it is also treated as an `enum`.

File: fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java [hunk 6]
```diff
@@ -326,8 +339,23 @@ private static ImmutableMap<Long, ComputeNode> buildComputeNodeInfo(SystemInfoSe
                 }
                 computeNodes.put(computeNode.getId(), computeNode);
             }
-            return ImmutableMap.copyOf(computeNodes);
+        } else { //numUsedComputeNodes >= idToComputeNode.size()
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/qe/scheduler/DefaultWorkerProvider.java:342]: This branch and the previous branch `else if (numUsedComputeNodes < idToComputeNode.size()) {` should be merged to one, and follow the same steps:
1. Get enough compute nodes from `idToComputeNode`, until `computeNodes.size() < numUsedComputeNodes`.
2. If the policy is `ALL_NODES`, then continually get enough compute nodes from `idToBackend`, until `computeNodes.size() < numUsedComputeNodes`.

Otherwise, the result `computeNodes` may contain some non-available compute nodes, but `computeNodes.size() == numUsedComputeNodes`.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 5 above)
Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:805]: ok，i will update later

Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: @kevincai @ZiheLiu i have updated the pr according to your suggestion , please approve the workflow and review again.

**Review by ZiheLiu — DISMISSED** [review]

**Comment by MatthewH00** [comment]
@kevincai Could please check the [CI PIPELINE / FE UT]，it faild , like hive env problem from log

**Comment by MatthewH00** [comment]
@kevincai it failed in part of AnalyzeSetVariableTest.java in FE UT . but i look i would fail .  so i adjust the test demo to remove the variable analyzeFail check (see the latest commit for detail). could you please approve the workflow again ?

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java (hunk 5 above)
Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java]: @kevincai @ZiheLiu It failed in part of AnalyzeSetVariableTest.java in FE UT. Becase it not contain IllegalArgumentException in the analyzeFail function. i add it. Could you please approve the workflow and review again ?

**Comment by MatthewH00** [comment]
@kevincai Could you please approve the workflow again? i fix a problem(It failed in part of AnalyzeSetVariableTest.java in FE UT. Becase it not contain IllegalArgumentException in the analyzeFail function. i add it) yesterday night.

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 7]
```diff
@@ -1935,7 +1942,7 @@ public SessionVariable setHiveTempStagingDir(String hiveTempStagingDir) {
 
     public SessionVariableConstants.ChooseInstancesMode getChooseExecuteInstancesMode() {
         return Enums.getIfPresent(SessionVariableConstants.ChooseInstancesMode.class,
-                        StringUtils.upperCase(chooseExecuteInstancesMode))
+                StringUtils.upperCase(chooseExecuteInstancesMode))
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:1945]: unrelated change.

File: fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeSetVariableTest.java [hunk 8]
```diff
@@ -271,4 +271,18 @@ public void testSetAdaptiveDopMaxBlockRowsPerDriverSeq() {
...
     }
+
+    @Test
+    public void testComputationFragmentSchedulingPolicy() {
+        String sql;
+
+        sql = "SET computation_fragment_scheduling_policy = compute_nodes_only";
+        analyzeSuccess(sql);
+
+        sql = "SET computation_fragment_scheduling_policy = all_nodes";
```
Comment (kevincai) [inline fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeSetVariableTest.java:282]: add mixed upper/lower case, ensure the variable value is case-insensitive.

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java [hunk 9]
```diff
@@ -211,6 +211,16 @@ private static void analyzeSystemVariable(SystemVariable var) {
             }
         }
 
+        if (variable.equalsIgnoreCase(SessionVariable.COMPUTATION_FRAGMENT_SCHEDULING_POLICY)) {
+            SessionVariableConstants.ComputationFragmentSchedulingPolicy computationFragmentSchedulingPolicy =
+                    Enums.getIfPresent(SessionVariableConstants.ComputationFragmentSchedulingPolicy.class,
+                            StringUtils.upperCase(resolvedExpression.getStringValue())).orNull();
+            if (computationFragmentSchedulingPolicy == null) {
+                String legalValues = Joiner.on(" | ").join(SessionVariableConstants.ComputationFragmentSchedulingPolicy.values());
+                throw new IllegalArgumentException("Legal values of computation_fragment_scheduling_policy are " + legalValues);
```
Comment (kevincai) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java:220]: maybe SematicException would be better?

Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java:220]: i refer to other variable like choose_execute_instances_mode , use the IllegalArgumentException . need i to update to SematicException ?

Comment (kevincai) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java:220]: most of the exceptions in this file are `SemanticException` for invalid input value. `choose_execute_instances_mode` is an exception. Thought it would be good to keep the same as the majority?

Comment (MatthewH00) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/SetStmtAnalyzer.java]: i have updated it, please check and approve workflow again.

**Review by kevincai — APPROVED** [review]

**Review by ZiheLiu — APPROVED** [review]

**Comment by MatthewH00** [comment]
@kevincai It looks like need [StarRocks/analyzer] to review also, could you please help to look for related classmate to review ?

**Review by Seaven — APPROVED** [review]
